hail/src/main/scala/is/hail/rvd/AbstractRVDSpec.scala

```diff
@@ -484,48 +481,76 @@ case class IndexedRVDSpec2(
   ): IR => TableStage = newPartitioner match {
     case Some(np) =>
       val part = partitioner(ctx.stateManager)
+      /* ensure the old and new partitioners have the same key, and ensure the new partitioner is
+       * strict */
       val extendedNP = np.extendKey(part.kType)
-      val tmpPartitioner = part.intersect(extendedNP)
 
       assert(key.nonEmpty)
 
-      val rSpec = typedCodecSpec
-      val reader =
-        ir.PartitionNativeReaderIndexed(rSpec, indexSpec, part.kType.fieldNames, uidFieldName)
-
-      val absPath = path
-      val partPaths = tmpPartitioner.rangeBounds.map(b => partFiles(part.lowerBoundInterval(b)))
-
-      val kSize = part.kType.size
-      absolutePartPaths(path)
-      assert(tmpPartitioner.rangeBounds.size == partPaths.length)
-      val contextsValues: IndexedSeq[Row] = tmpPartitioner.rangeBounds.map { interval =>
-        val partIdx = part.lowerBoundInterval(interval)
-        val partPath = partFiles(partIdx)
+      val reader = ir.PartitionNativeReaderIndexed(
+        typedCodecSpec,
+        indexSpec,
+        part.kType.fieldNames,
+        uidFieldName,
+      )
+
+      def makeCtx(oldPartIdx: Int, newPartIdx: Int): Row = {
+        val oldInterval = part.rangeBounds(oldPartIdx)
+        val partFile = partFiles(oldPartIdx)
+        val intersectionInterval =
+          extendedNP.rangeBounds(newPartIdx)
+            .intersect(extendedNP.kord, oldInterval).get
         Row(
-          partIdx.toLong,
-          s"$absPath/parts/$partPath",
-          s"$absPath/${indexSpec.relPath}/$partPath.idx",
-          RVDPartitioner.intervalToIRRepresentation(interval, kSize),
+          oldPartIdx.toLong,
+          s"$path/parts/$partFile",
+          s"$path/${indexSpec.relPath}/$partFile.idx",
+          RVDPartitioner.intervalToIRRepresentation(intersectionInterval, part.kType.size),
         )
       }
 
-      assert(TArray(reader.contextType).typeCheck(contextsValues))
+      val (nestedContexts, newPartitioner) = if (filterIntervals) {
```

Comment:
Trying to get my head around why these code paths are subtly different and both necessary.
What's the meaning of `filterIntervals`? Does it mean something like drop partition reads of an empty interval?